smart-contracts/contracts/cl-vault/src/contract.rs

```diff
@@ -174,9 +179,62 @@ pub fn reply(deps: DepsMut, env: Env, msg: Reply) -> Result<Response, ContractEr
     }
 }
 
-#[entry_point]
-pub fn migrate(_deps: DepsMut, _env: Env, _msg: MigrateMsg) -> Result<Response, ContractError> {
-    Ok(Response::new().add_attribute("migrate", "successful"))
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn migrate(deps: DepsMut, env: Env, _msg: MigrateMsg) -> Result<Response, ContractError> {
+    let vault_denom = VAULT_DENOM.load(deps.storage)?;
+
+    let mut total_old = Uint256::zero();
+    // let mut total_new = Uint128::zero();
+
+    let mut response = Response::new();
+
+    let vals: Result<Vec<(Addr, Uint128)>, ContractError> = SHARES
+        .range(deps.storage, None, None, cosmwasm_std::Order::Ascending)
+        .map(|val| -> Result<(Addr, Uint128), ContractError> {
+            let (user, shares) = val?;
+            // Convert Uint128 to Uint256 for old shares
+            let shares_256 = Uint256::from(shares.u128());
+
+            // Perform the division
+            let new_shares_256 = shares_256
+                .checked_div(Uint256::from_u128(10u128).pow(18))
+                .expect("Underflow");
+
+            // Convert back to Uint128 for new shares, handling potential overflow
+            let new_shares: Uint128 = Uint128::try_from(new_shares_256)
+                .expect("Conversion from Uint256 to Uint128 failed due to overflow");
+
+            total_old = total_old.checked_add(shares_256)?;
+            Ok((user, new_shares))
+        })
+        .collect();
+
+    for (user, new_shares) in vals? {
+        SHARES.save(deps.storage, user, &new_shares)?;
```

Comment:
If this save is causing borrow checker issues on deps as mut/immutable, I would prefer to do the save in a seperate loop so we dont have to recalculate old_shares_256